File: test/error-bound/algorithms.ts
/**
 * @file Pre-test for the well known EFTs and algorithms
 *
 * Generates inputs and collect outputs for each algorithm then export the whole
 * as JSON so we can import it and do proper tests in Julia using BigFloat.
 *
 * Derived algorithms that are just specialization of those ones are tested
 * separately).
 */

import {
  f64,
  TwoF64,
  F64_SPLITTER,
  normalize as _normalize,
  split,
  twoSum,
  twoProd,
} from '../../src/index';

import { exponent, FLOAT64_MIN } from '@lvlte/ulp';
import fs from 'node:fs';

import {
  DWPlusFP,
  AccurateDWPlusDW,
  DWTimesFP1,
  DWTimesDW1,
  DWDivFP3,
  DWDivDW2,
} from '../../src/base/algorithms';

import {
  UnionToIntersection,
  Expand,
  randomFn,
  pairsInRange,
  signCombinations
} from '../utils';

interface FnSig {
  'op1': (x: f64) => TwoF64;
  'op2': (x: TwoF64) => TwoF64;
  'op11': (x: f64, y: f64) => TwoF64;
  'op21': (x: TwoF64, y: f64) => TwoF64;
  'op22': (x: TwoF64, y: TwoF64) => TwoF64;
}

// Wrap normalize so it is tested with the |x| ≥ |y| condition satisfied
const normalize: typeof _normalize = (x, y) => {
  return Math.abs(x) >= Math.abs(y) ? _normalize(x, y) : _normalize(y, x);
};

// Functions to test grouped by signature
const fnBySig = {
  'op1': {split},
  'op11': {normalize, twoSum, twoProd},
  'op21': {DWPlusFP, DWTimesFP1, DWDivFP3},
  'op22': {AccurateDWPlusDW, DWTimesDW1, DWDivDW2},
} satisfies Partial<{
  [K in keyof FnSig]: { [fnName: string]: FnSig[K] }
}>;

type FnBySig = typeof fnBySig;
type TestedFunctions = UnionToIntersection<FnBySig[keyof FnBySig]>;
type FnName = keyof TestedFunctions;

type ArgsListBySig = { [K in keyof FnBySig]: Parameters<FnSig[K]>[] };
type FnOutputList = { [K in FnName]: ReturnType<TestedFunctions[K]>[] }

// Pseudo-random number generator
const SEED = Math.sqrt(2);
const random = randomFn(SEED, true);

// Lists of arguments (grouped by FnSig) to pass to the TestedFunctions
const argsList: ArgsListBySig = {
  'op1': [], 'op11': [], 'op21': [], 'op22': []
};

// split is not immune to overflow
const E_SPLIT_MAX = exponent(Number.MAX_VALUE/F64_SPLITTER);

// Exponent range for the generated numbers (roughly the largest window for
// which the error bounds can be checked properly since they assume no overflow
// /underflow occurs during calculations).
const e_shift = 0; // decrease to shift the window towards subnormals
const emin = Math.floor(exponent(FLOAT64_MIN)/2) + e_shift;
const emax = Math.min(0, emin) + E_SPLIT_MAX;

// Fill argsList with number combinations in the domain [±2^emin, ±2^emax]
for (const [e1, e2] of pairsInRange(emin, emax, 3)) {
  for (const [s1, s2] of signCombinations) {
    const w = random(e1, s1);
    const x = random(e1, s1);
    const y = random(e2, s2);
    const z = random(e2, s2);

    const wx = normalize(w, x);
    const yz = normalize(y, z);

    if ([w, x, y, z, ...wx, ...yz].some(v => !Number.isFinite(v))) {
      continue;
    }

    argsList['op1'].push([x]);
    argsList['op11'].push([x, y]);
    argsList['op21'].push([wx, z]);
    argsList['op22'].push([wx, yz]);
  }
}

// Produce the list of outputs keyed by function given argsList
const fnOutput = {} as FnOutputList;
for (const sid in fnBySig) {
  const fnGroup = fnBySig[sid as keyof FnBySig];
  const argsGroup = argsList[sid as keyof FnBySig]
  for (const fnName in fnGroup) {
    fnOutput[fnName as FnName] = [];
    const fn = (fnGroup as TestedFunctions)[fnName as FnName];
    const fnOut = fnOutput[fnName as FnName]!;
    for (const args of argsGroup) {
      // @ts-ignore (TS doesn't understand correlated unions)
      const result = fn(...args);
      fnOut.push(result);
    }
  }
}

// Inputs/Outputs object
const testset = { argsList, fnOutput} as Expand<{
  argsList: ArgsListBySig, fnOutput: FnOutputList
}>;

// Export as JSON
const testsetJSON = JSON.stringify(testset);
fs.writeFileSync('test/error-bound/algorithms-testset.json', testsetJSON, 'utf8');

File: test/error-bound/algorithms.jl
# Test the accuracy of twofloat's main algorithms
# Relative error bounds: cf. paper from J.M. Muller et al.
# Function inputs/outputs dataset is created by /test/algorithms.ts

json = read("$(@__DIR__)/testset/algorithms.json", String)
testset = JSON.parse(json, TestSet; null=NaN)

args_list = testset.argsList
fn_output = testset.fnOutput

coverage = Dict(keys(fn_output) .=> false)
overflow = Dict(keys(fn_output) .=> 0)

println()
@testset verbose = true "Error-Free Transforms ────────────────" begin #########

    @testset "split" begin
        _test(Dict(
            "fn" => "split",
            "args" => args_list.op1,
            "rel_err_bound" => 0,
            "compute" => x -> big(x)
        ))
    end

    @testset "normalize (fast2Sum, fast2Diff)" begin
        _test(Dict(
            "fn" => "normalize",
            "args" => args_list.op11,
            "rel_err_bound" => 0,
            "compute" => (x, y) -> big(x) + big(y)
        ))
    end

    @testset "twoSum (twoDiff, add11, sub11)" begin
        _test(Dict(
            "fn" => "twoSum",
            "args" => args_list.op11,
            "rel_err_bound" => 0,
            "compute" => (x, y) -> big(x) + big(y)
        ))
    end

    @testset "twoProd (mul11, twoSquare, square_1)" begin
        _test(Dict(
            "fn" => "twoProd",
            "args" => args_list.op11,
            "rel_err_bound" => 0,
            "compute" => (x, y) -> big(x) * big(y)
        ))
    end
end

println()
@testset verbose = true "Error bounds op21 ────────────────────" begin #########

    @testset "DWPlusFP (add21, sub21)" begin
        _test(Dict(
            "fn" => "DWPlusFP",
            "args" => args_list.op21,
            "rel_err_bound" => 2u^2,
            "compute" => ((xhi, xlo), y) -> big(y) + big(xhi) + big(xlo)
        ))
    end

    @testset "DWTimesFP1 (mul21)" begin
        _test(Dict(
            "fn" => "DWTimesFP1",
            "args" => args_list.op21,
            "rel_err_bound" => 1.5u^2 + 4u^3,
            "compute" => ((xhi, xlo), y) -> big(xhi)*big(y) + big(xlo)*big(y)
        ))
    end
    @testset "DWDivFP3 (div11, div21, inv1)" begin
        _test(Dict(
            "fn" => "DWDivFP3",
            "args" => args_list.op21,
            "rel_err_bound" => 3u^2,
            "compute" => ((xhi, xlo), y) -> big(xhi)/big(y) + big(xlo)/big(y)
        ))
    end
end

println()
@testset verbose = true "Error bounds op22 ────────────────────" begin #########

    @testset "AccurateDWPlusDW (add22, sub22)" begin
        _test(Dict(
            "fn" => "AccurateDWPlusDW",
            "args" => args_list.op22,
            "rel_err_bound" => 3u^2 + 13u^3,
            "compute" => ((xhi,xlo),(yhi,ylo)) -> big(xhi) + big(yhi) + big(xlo) + big(ylo)
        ))
    end

    @testset "DWTimesDW1 (mul22, square_2)" begin
        _test(Dict(
            "fn" => "DWTimesDW1",
            "args" => args_list.op22,
            "rel_err_bound" => 5u^2,
            "compute" => ((xhi,xlo),(yhi,ylo)) -> (big(xhi) + big(xlo)) * (big(yhi) + big(ylo))
        ))
    end

    @testset " (div22, inv2)" begin
        _test(Dict(
            "fn" => "DWDivDW2",
            "args" => args_list.op22,
            "rel_err_bound" => 15u^2 + 56u^3,
            "compute" => ((xhi,xlo),(yhi,ylo)) -> (big(xhi) + big(xlo)) / (big(yhi) + big(ylo))
        ))
    end
end

###

println()
@testset "Algorithm coverage ───────────────────" begin ########################
    for (fn, covered) in coverage
        @test (fn, covered) == (fn, true)
    end
end

overflowed = filter(((fn , ov_count),) -> ov_count > 0, overflow)
if !isempty(overflowed)
    println()
    @info ["overflow\n ", (rpad(k, 20, ' ') * "$v\n " for (k, v) in overflowed)...] |> join
end
